Script that creates web-platform-test wrappers for test262's tests

dpino opened the pull request:
The scripts reads out 'test262/test' and generates web-platform-test wrappers for each test. The generated tests are stored at 'js/test262'.

Each generated web-platform-test runs the target test262 test in 4 different agents: IFrame, Window, DedicatedWorker and SharedWorker. For each agent usually two tests are generated: one in strict-mode and another one in non-strict mode (unless the test indicates otherwise).

For IFrame and Window agents an HTML file is created on-the-fly containing the target test. In the case of DedicatedWorker and SharedWorker a JavaScript test is created on-the-fly containing the target test. The harnessing code runs the test and checks whether there was any error.

Currently when running the generated web-platform-test wrappers many tests fail. More details at issue web-platform-tests/rfcs#230

I post this PR with no intention of merging it yet, but to gather feedback and discuss what needs to be done from here to get the PR landed.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/w3c/web-platform-tests/8980)
<!-- Reviewable:end -->

dpino committed:
Script that creates web-platform-test wrappers for test262's tests

The generated tests run the target test262 within different agent.
Currently supported agents are: IFrame, Window, DedicatedWorker and
SharedWorker. For each agent two tests are generated, one in strict mode
and another in non-strict mode, unless the test indicates otherwise.

The generated tests are placed at js/test262/.

dpino commented:
How to run the PR:

```bash
$ ./test262-to-wpt test262
Generating web-platform-test wrappers for test262
Output: js/test262
```

Now run one of the generated WPT tests, for instance:

```
$ ./wpt run firefox js/test262/intl402/Date/prototype/toLocaleDateString/13.3.2_L15.html
web-platform-test
~~~~~~~~~~~~~~~~~
Ran 9 checks (1 tests, 8 subtests)
Expected results: 9
OK
```

Helper script to run all the tests and save output to file. Since several tests fail the tests are run with _no-restart-on-unexpected_ flag on, to speed up overall running time.

**run-all.sh**

```
#!/usr/bin/env bash

FOLDERS=(
   annexB
   built-ins
   harness
   intl402
   language
)

for each in ${FOLDERS[@]}; do
   time ./wpt run --no-restart-on-unexpected firefox "js/test262/test/${each}" | tee "${each}.output"
done
```

dpino commented:
Some considerations regarding the PR:

* The generated web-platform-tests depend on test262's harness. Test262's harness code is expected to be found at folder 'resources/test262/harness'. This code is not part of the PR. Perhaps it should be copied when the tests are generated.
* Should the generated tests be part of web-platform-test codebase? If so it would be necessary to discuss an updating process. Perhaps by maintaining a file with a timestamp for the original tests and generated test. Ideally generated test should not be manually modified.
* Supporting ServiceWorkers is tricky as they cannot be generate on-the-fly with a Blob. Probably support can be added if the ServiceWorker test is generated beforehand along with the WPT generation process.
* 262 tests that use the JavaScript Shell Agent API should be treated in a special manner, covering all the possible scenarios for the agents supported (See https://github.com/tc39/test262/issues/1351).

w3c-bots commented:
# Build ERRORED

Started: 2018-01-10 17:05:40
Finished: 2018-01-10 17:13:52


<h2>Failing Jobs</h2>
<ul>

<li>lint</li>

<li>safari:11.0</li>

<li>MicrosoftEdge:14.14393</li>

</ul>




View more information about this build on:

- [WPT PR Status](http://pulls.web-platform-tests.org/build/22590)
- [TravisCI](https://travis-ci.org/w3c/web-platform-tests/builds/327298211)

rwaldron commented on resources/test262-harness.js:
```diff
@@ -0,0 +1,39 @@
+    return global.$262;
+}
+
+// TODO: $DONE is used by Promise tests and others, but I don't know from where to fetch it.
```
`$DONE` is added by `test262-compiler`: https://github.com/bterlson/test262-compiler/blob/master/lib/compilerdeps.js, but the plan is to eliminate the `test262-compiler` dependency by consolidating the "compiler" operations and boilerplate into `test262-stream` (https://github.com/bocoup/test262-stream/pull/7). Once that's complete, `test262-harness` is going to be updated to use `test262-stream`.

rwaldron commented:
I recommend blacklisting the **test262/test/harness** directory, this is just tests for test262's harness itself and is of no value outside of test262.

rwaldron commented on resources/test262-agent-harness.js:
```diff
@@ -0,0 +1,335 @@
+    test = "'use strict';\n" + test;
+  }
+  if (attrs.type == 'SyntaxError' && attrs.phase == 'runtime') {
+      // If phase is 'runtime' the error will be caught here. If phase is 'early' the
```
`negative:` phases are more specific in the latest version (2.0.0) of https://github.com/tc39/test262/blob/master/INTERPRETING.md, see [negative](https://github.com/tc39/test262/blob/master/INTERPRETING.md#negative) for more details, but the short version is that `early` is now `parse`.

rwaldron replied:
You can safely check for both for the time being.

rwaldron commented on resources/test262-agent-harness.js:
```diff
@@ -0,0 +1,335 @@
+      // error will be handled at the window.onerror event.
+      output.push('try { eval("');
+      output.push(escapeEvalExpr(test));
+      output.push('"); }\ncatch (e) { assert.sameValue(e instanceof SyntaxError, true); }');
```
The condition above will match tests containing the following:

```
negative:
  phase: runtime
  type: SyntaxError
```

The only tests that contain this particular set of attributes are those that test module code, direct eval and indirect eval: 

- [eval-code/direct/var-env-global-lex-non-strict.js](https://github.com/tc39/test262/tree/master/test/language/eval-code/direct/var-env-global-lex-non-strict.js)
- [eval-code/indirect/parse-failure-2.js](https://github.com/tc39/test262/tree/master/test/language/eval-code/indirect/parse-failure-2.js)
- [module-code/instn-iee-err-ambiguous-as.js](https://github.com/tc39/test262/tree/master/test/language/module-code/instn-iee-err-ambiguous-as.js)
- [module-code/instn-iee-err-ambiguous.js](https://github.com/tc39/test262/tree/master/test/language/module-code/instn-iee-err-ambiguous.js)
- [module-code/instn-iee-err-circular-as.js](https://github.com/tc39/test262/tree/master/test/language/module-code/instn-iee-err-circular-as.js)
- [module-code/instn-iee-err-circular.js](https://github.com/tc39/test262/tree/master/test/language/module-code/instn-iee-err-circular.js)
- [module-code/instn-iee-err-dflt-thru-star-as.js](https://github.com/tc39/test262/tree/master/test/language/module-code/instn-iee-err-dflt-thru-star-as.js)
- [module-code/instn-iee-err-dflt-thru-star.js](https://github.com/tc39/test262/tree/master/test/language/module-code/instn-iee-err-dflt-thru-star.js)
- [module-code/instn-iee-err-not-found-as.js](https://github.com/tc39/test262/tree/master/test/language/module-code/instn-iee-err-not-found-as.js)
- [module-code/instn-iee-err-not-found.js](https://github.com/tc39/test262/tree/master/test/language/module-code/instn-iee-err-not-found.js)
- [module-code/instn-named-err-ambiguous-as.js](https://github.com/tc39/test262/tree/master/test/language/module-code/instn-named-err-ambiguous-as.js)
- [module-code/instn-named-err-ambiguous.js](https://github.com/tc39/test262/tree/master/test/language/module-code/instn-named-err-ambiguous.js)
- [module-code/instn-named-err-dflt-thru-star-as.js](https://github.com/tc39/test262/tree/master/test/language/module-code/instn-named-err-dflt-thru-star-as.js)
- [module-code/instn-named-err-dflt-thru-star-dflt.js](https://github.com/tc39/test262/tree/master/test/language/module-code/instn-named-err-dflt-thru-star-dflt.js)
- [module-code/instn-named-err-not-found-as.js](https://github.com/tc39/test262/tree/master/test/language/module-code/instn-named-err-not-found-as.js)
- [module-code/instn-named-err-not-found-dflt.js](https://github.com/tc39/test262/tree/master/test/language/module-code/instn-named-err-not-found-dflt.js)
- [module-code/instn-named-err-not-found.js](https://github.com/tc39/test262/tree/master/test/language/module-code/instn-named-err-not-found.js)
- [module-code/instn-star-err-not-found.js](https://github.com/tc39/test262/tree/master/test/language/module-code/instn-star-err-not-found.js)

These can't be nested in a `try { eval(...)` }...`

rwaldron replied:
Here's a better illustration from firefox nightly:

This is code exactly as `prepareTest(...)` returns it: 

```js
try { eval("import x from './instn-named-err-not-found-empty_FIXTURE.js';"); }
catch (e) { assert.sameValue(e instanceof SyntaxError, true); }
```

If I modify it to:

```js
try { eval("import x from './instn-named-err-not-found-empty_FIXTURE.js';"); }
catch (e) { console.log(e); }
```

And put it in an html file, with the same `<script type="text/javascript">` (this is important) : 

```html
<!DOCTYPE html>
<script type="text/javascript">
  try { eval("import x from './instn-named-err-not-found-empty_FIXTURE.js';"); }
  catch (e) { console.log(e); }
</script>
```

This will log: 

```
SyntaxError: import declarations may only appear at top level of a module
```


If I change that to (`<script type="module">` is the change): 

```js
<!DOCTYPE html>
<script type="module">
  try { eval("import x from './instn-named-err-not-found-empty_FIXTURE.js';"); }
  catch (e) { console.log(e); }
</script>
```

...The same error occurs. Also, in neither case does a network request get made for `instn-named-err-not-found-empty_FIXTURE.js`.


Finally, when I run this: 

```html
<!DOCTYPE html>
<script type="module">
import x from './instn-named-err-not-found-empty_FIXTURE.js';
</script>
```

The outcome is: 


```
SyntaxError: import not found: default
```

Which is the `SyntaxError` this test was written to verify. There is also a network request made for `instn-named-err-not-found-empty_FIXTURE.js`. (I did this whole exercise in Firefox Nightly with `dom.moduleScripts.enabled` set to `true` and Safari Technology Preview. 


------------------------------------------


Note: I had to manually make a file called `instn-named-err-not-found-empty_FIXTURE.js`, because the real `instn-named-err-not-found-empty_FIXTURE.js` has been turned into a file called `instn-named-err-not-found-empty_FIXTURE.html`, with all of the same treatments. Easily fixed by blacklisting `_FIXTURE.js` files.

dpino replied:
Nice catch! I will rework the code as suggested.

rwaldron commented on resources/test262-agent-harness.js:
```diff
@@ -0,0 +1,335 @@
+
+  <!-- Test262 required libraries -->
+  <script src="http://localhost:8000/resources/test262/harness/assert.js"><\/script>
+  <script src="http://localhost:8000/resources/test262/harness/sta.js"><\/script>
```
This domain doesn't work if the tests are run from `http://web-platform.test:8000` (in fact, my localhost:8000 is not the same as http://web-platform.test:8000)

jgraham replied:
Add `.sub.` to the filename and then you can use substitutions like `{{host}}` c.f. the rather outdated documentation at https://wptserve.readthedocs.io/en/latest/pipes.html#sub

jgraham replied:
(or use just paths, of course, in this case, but in general the feature is there if you need it).

rwaldron replied:
@jgraham to clarify, do you mean, eg:

```html
<script src="http://{{host}}:{{ports[http][0]}}/resources/test262/harness/sta.js"><\/script>
```

dpino replied:
Thanks for the hint. I wasn't happy with hardcoding the host and the port but I couldn't find other way to make it work. I'll fix.

rwaldron commented on test262-to-wpt:
```diff
@@ -0,0 +1,404 @@
+  <head>
+    <meta charset="utf-8">
+    <title>###TITLE###</title>
+    <meta name="help" href="https://storage.spec.whatwg.org/#dom-storagemanager-persisted">
```
I don't think you meant to include this?

dpino replied:
Right, likely a copy-paste error.

rwaldron commented on resources/test262-harness.js:
```diff
@@ -0,0 +1,39 @@
+			if (typeof postMessage !== 'function') {
+				throw new Error('No method available to detach an ArrayBuffer');
+			} else {
+                postMessage(null, '*', [buffer]);
```
Strange indentation 0_o

dpino replied:
It seems there's a mix up of tabs and spaces in that file. I will review all the files and make sure there are only spaces.

dpino committed:
Remove tabs

dpino committed:
Remove unneeded meta tags

dpino committed:
Include test262-harness at global level

dpino committed:
Parametrized host and port

dpino committed:
Blacklist harness and copy _FIXTURE.js files

rwaldron commented:
I know that @jugglinmike and @jgraham have discussed this extensively in the past and I want to make sure that all of their interests/concerns are heard and/or addressed.

dpino committed:
Use Python's YAML and JSON libraries

dpino committed:
Simplify SyntaxError handling

If a test expects a SyntaxError, regardless of the value of phase, the
error is always managed in window.onerror.

dpino committed:
Refactor test262_as_html

dpino committed:
Include test262-harness.js

dpino committed:
Redefine  and avoid calling installAPI if test is a Worker

dpino commented:
@rwaldron I've addressed all the comments in the review. Regarding the definition of $DONE I made it rethrow an error in case one is passed as argument. I also changed how errors are handled. I caught all types of errors now (regardless what phase) at the window.onerror event PTAL.

I will update the PR soon with the tests results.

dpino committed:
Return global.62

dpino committed:
Remove trailing whitespaces

dpino committed:
Remove custom tojson function

dpino committed:
Replace carriage-return

Replace CR + LF for LF only.
Replace CR for LF.

dpino committed:
ES Shell is not supported in a Worker context

Tests that use ES functions such as createRealm, evalScript or
dettachArrayBuffer are not supported in a Worker context.

These operations are implemented in a Window context by appending new
elements to the DOM (an iframe in the case of createRealm and a script
element in the case of evalScript). Workers don't have access to the
Window object. On the other hand, creating a new realm by instantiating
a new Worker cannot be a solution. The new realm (a Worker) won't
provide the same functions as a Window, which what the tests expect to find
after calling createRealm (for instance $262.createRealm().eval(...))

dpino committed:
Return MIME type 'module' if flags includes 'module'

dpino committed:
Implement IsHTMLDDA

dpino committed:
Tests marked with 'raw' flag are only executed in non-strict mode

dpino committed:
Properly handling of async tests

Tests marked with flag 'async' should include doneprintHandle.js.
Implement ES Shell function 'print' that receives a message
'Test262:AsyncTestComplete' in case the test completed successfully.
Otherwise, throw error. If print is never called, rely on WPT timeout
mechanism.

dpino committed:
Properly handling of negative tests

Tests with a negative property are supposed to throw an Exception. The
thrown exception should have the same type as negative.type.

dpino committed:
Reintroduce handling of early errors

Syntax errors, the type that are emitted during parse or early phase,
cab be caught if the test is wrapped within a try/catch.

dpino committed:
Caught early tests on SyntaxError too

For many tests Firefox throws a SyntaxError instead of a ReferenceError.

dpino committed:
Improve module tests results

Many module-code tests fail cause they try to import or export a JavaScript
module which path is no longer valid. It's necesary to replace the path with
an URL. This change implies to activate CORS in the WPT WebServer.

dpino committed:
Generate tests for any path

dpino committed:
IsHTMLDDA should return null if argument is null or empty string

dpino committed:
Fix run_in_eval

dpino committed:
global should be the global object

dpino commented:
I have pushed some more commits that help fixing several tests.

The most important change it's fixing tests that are related with modules. The tests complain they could not locate the module, since the path was broken. The following error was returned:

```
SyntaxError: import declarations may only appear at top level of a module
```

I tried to fix the path by using relative and absolute paths but it only worked if I replaced the original path with an URL. This led me to activate CORS in the WPT WebServer plus hardcoding "host" and "port" in the URL (which I know it's wrong). I tried to encode the URL using placeholders and passing ?pipe=sub to the URL that points to the target file containing the test, but it didn't work in this case (does this mechanism work with modules import too?).

foolip commented:
@dpino what is the status of this? I see this has been stale for almost two years, but if you do want to keep working on it, I think it'd be appropriate to put it through the RFC process at https://github.com/web-platform-tests/rfcs.

There are also other external test suites that might be interesting to run as part of wpt, or at least get results for into wpt.fyi, see https://github.com/web-platform-tests/wpt/issues/19438 for a suggested experiment for Gecko's mochitests.

annevk commented:
This was an experiment funded by Mozilla to see if there's value in running JavaScript tests with different types of agents. We didn't proceed eventually as it seems there doesn't appear to be a lot of value to be gained and properly abstracting everything would be quite a bit of effort.

foolip commented:
I see, in that case I guess this PR should be closed?

annevk commented:
I think that's reasonable, but I'll let @dpino make the call.

foolip commented:
No reply from @dpino, so I'm closing.
